### PR TITLE
Add rosdoc2 documentation job type

### PR DESCRIPTION
**Pull request by clalancette**
This PR allows us to start using https://github.com/ros-infrastructure/rosdoc2 to generate per-package API documentation for packages on the buildfarm.  In order to make this truly work, we'll need a ros_buildfarm_config file like https://github.com/ros-staging/ros_buildfarm_config/blob/848d5107f7e4f3925b2ad08786784387b53a9cf2/rolling/doc-build.yaml on the main buildfarm.

As it stands, these jobs have a few limitations (some of which come from this code, some of which come from rosdoc2):
    
1.  This will really only work for C++ packages.  To make it work for Python requires an update to the rosdoc2 code.
2.  It will only generate documentation against sources.  To make it look at build or install artifacts as well, we'll need to update both rosdoc2 and this code.
3.  It will not do any cross-referencing outside of the current repository.  To make this work requires generating, storing, and retrieving a cross-reference database.

Nonetheless, I think this is a good first step and will allow us to at least build basic API documentation for packages.

You can see an example list of jobs that would be created by this PR at https://build.test.ros2.org/view/Rdoc/ .

To truly host the documentation we'll also need to properly setup the Apache configuration at https://docs.ros.org , but that is somewhat orthogonal to this.

Note that the first 2 commits in this series are bugfixes, so can be applied with or without the third one.

Part of https://github.com/ros2/ros2/issues/1009

### CODE REVIEW

**Comment by nuclearsandwich** [comment]
>     2\. To make it look at build or install artifacts as well, we'll need to update both rosdoc2 and this code.

I'm pretty sure that `rcutils` documentation is basically broken by this but will it be sufficient for `rcl` and `rclcpp` where having full API documentation out the door would be a really great launch feature.

**Comment by clalancette** [comment]
> I'm pretty sure that `rcutils` documentation is basically broken by this but will it be sufficient for `rcl` and `rclcpp` where having full API documentation out the door would be a really great launch feature.

Can you explain a bit more why you think this won't work for rcutils?  In my local testing it seemed to do more-or-less what I wanted, but there may be something I'm missing.

**Comment by nuclearsandwich** [comment]
> Can you explain a bit more why you think this won't work for rcutils? In my local testing it seemed to do more-or-less what I wanted, but there may be something I'm missing.

Mostly this is based on memory so things may have changed but when building documentation ad-hoc we had to build in order to get documentation for generated code https://github.com/ros2/docs.ros2.org/issues/24

**Comment by clalancette** [comment]
> Mostly this is based on memory so things may have changed but when building documentation ad-hoc we had to build in order to get documentation for generated code [ros2/docs.ros2.org#24](https://github.com/ros2/docs.ros2.org/issues/24)

Oh yeah, that's right.  We won't be documenting anything that is generated (this goes for the RCLCPP macros as well).  But we should get everything else, and doing the building is a follow-on task that this framework should allow us to do.

**Comment by nuclearsandwich** [comment]
Did we discuss previously whether creating a separate rosdoc2 job template was the recommended course of action. I diffed the two and the differences are almost entirely in parameters to snippet functions and not in overall structure thus I'm torn between my general dislike of branching logic in template files and concern that creating a second template for rosdoc2 jobs will create work for us as we will have to update both job templates to make doc job changes going forward. I'm very interested in @cottsay's thoughts on this matter as he's spent more time in ros_buildfarm than I have and his work adding rpm support will have given him a much more developed stance than I'll have.


<details>
<summary>doc_job.xml.em..rosdoc2_job.xml.em</summary>

```diff
--- ros_buildfarm/templates/doc/doc_job.xml.em  2021-01-28 09:18:08.637066359 -0800
+++ ros_buildfarm/templates/doc/rosdoc2_job.xml.em      2021-05-28 07:39:59.991229169 -0700
@@ -42,7 +42,7 @@
         {
             'type': 'boolean',
             'name': 'skip_cleanup',
-            'description': 'Skip cleanup of catkin build artifacts as well as rosdoc index',
+            'description': 'Skip cleanup of build artifacts',
         },
     ],
 ))@
@@ -90,42 +90,10 @@
 @(SNIPPET(
     'builder_shell',
     script='\n'.join([
-        'echo "# BEGIN SECTION: Clone rosdoc_lite"',
-        'rm -fr rosdoc_lite',
-        'python3 -u $WORKSPACE/ros_buildfarm/scripts/wrapper/git.py clone --depth 1 https://github.com/ros-infrastructure/rosdoc_lite.git rosdoc_lite',
-        'git -C rosdoc_lite log -n 1',
-        'rm -fr rosdoc_lite/doc',
-        'echo "# END SECTION"',
-        '',
-        'echo "# BEGIN SECTION: Clone catkin-sphinx"',
-        'rm -fr catkin-sphinx',
-        'python3 -u $WORKSPACE/ros_buildfarm/scripts/wrapper/git.py clone --depth 1 https://github.com/ros-infrastructure/catkin-sphinx catkin-sphinx',
-        'git -C catkin-sphinx log -n 1',
-        'echo "# END SECTION"',
-    ]),
-))@
-@(SNIPPET(
-    'builder_shell',
-    script='\n'.join([
-        'echo "# BEGIN SECTION: rsync (most of) the rosdoc_index to agent"',
-        'rm -fr rosdoc_index',
-        'mkdir rosdoc_index',
-        '# since rsync fails if the source folder does not exist we need to check it before',
-        'if ssh %s@%s stat %s/%s > /dev/null 2> /dev/null; then' % \
-          (upload_user, upload_host, upload_root.rstrip('/'), rosdistro_name),
-        '  rsync -e ssh --stats' +
-        ' --prune-empty-dirs --recursive' +
-        ' --include="*/"' +
-        ' --include="api/*/manifest.yaml"' +
-        ' --include="deps/*"' +
-        ' --include="hashes/%s"' % doc_repo_spec.name +
-        ' --include="locations/*"' +
-        ' --include="metapackage_deps/*"' +
-        ' --include="symbols/*"' +
-        ' --exclude="*"' +
-        ' %s@%s:%s/%s $WORKSPACE/rosdoc_index' % \
-          (upload_user, upload_host, upload_root.rstrip('/'), rosdistro_name),
-        'fi',
+        'echo "# BEGIN SECTION: Clone rosdoc2"',
+        'rm -fr rosdoc2',
+        'python3 -u $WORKSPACE/ros_buildfarm/scripts/wrapper/git.py clone --depth 1 https://github.com/ros-infrastructure/rosdoc2.git rosdoc2',
+        'git -C rosdoc2 log -n 1',
         'echo "# END SECTION"',
     ]),
 ))@
@@ -158,26 +126,18 @@
         'export TZ="%s"' % timezone,
         'export PYTHONPATH=$WORKSPACE/ros_buildfarm:$PYTHONPATH',
         'if [ "$force" = "true" ]; then FORCE_FLAG="--force"; fi',
-        'python3 -u $WORKSPACE/ros_buildfarm/scripts/doc/run_doc_job.py' +
-        ' ' + config_url +
-        ' --rosdistro-index-url ' + rosdistro_index_url +
-        ' ' + rosdistro_name +
-        ' ' + doc_build_name +
-        ' ' + doc_repo_spec.name +
+        'python3 -u $WORKSPACE/ros_buildfarm/scripts/doc/run_rosdoc2_job.py' +
         ' ' + os_name +
         ' ' + os_code_name +
         ' ' + arch +
-        ' --build-tool ' + build_tool +
-        ' --vcs-info "%s %s %s"' % (doc_repo_spec.type, doc_repo_spec.version if doc_repo_spec.version is not None else '', doc_repo_spec.url) +
         ' ' + ' '.join(repository_args) +
-        ' $FORCE_FLAG' +
         ' --dockerfile-dir $WORKSPACE/docker_generating_docker',
         'echo "# END SECTION"',
         '',
         'echo "# BEGIN SECTION: Build Dockerfile - generating doc task"',
         'cd $WORKSPACE/docker_generating_docker',
         'python3 -u $WORKSPACE/ros_buildfarm/scripts/misc/docker_pull_baseimage.py',
-        'docker build --force-rm -t doc_task_generation.%s_%s .' % (rosdistro_name, doc_repo_spec.name.lower()),
+        'docker build --force-rm -t rosdoc2_task_generation.%s_%s .' % (rosdistro_name, doc_repo_spec.name.lower()),
         'echo "# END SECTION"',
         '',
         'echo "# BEGIN SECTION: Run Dockerfile - generating doc task"',
@@ -187,17 +147,13 @@
         ' --rm ' +
         ' --cidfile=$WORKSPACE/docker_generating_docker/docker.cid' +
         ' -e=HOME=/home/buildfarm' +
-        ' -e=TRAVIS=$TRAVIS' +
-        ' -e=ROS_BUILDFARM_PULL_REQUEST_BRANCH=$ROS_BUILDFARM_PULL_REQUEST_BRANCH' +
         ' -v $WORKSPACE/ros_buildfarm:/tmp/ros_buildfarm:ro' +
-        ' -v $WORKSPACE/rosdoc_lite:/tmp/rosdoc_lite:ro' +
-        ' -v $WORKSPACE/catkin-sphinx:/tmp/catkin-sphinx:ro' +
-        ' -v $WORKSPACE/rosdoc_index:/tmp/rosdoc_index:ro' +
+        ' -v $WORKSPACE/rosdoc2:/tmp/rosdoc2:ro' +
         ' -v $WORKSPACE/ws:/tmp/ws' +
         ' -v $WORKSPACE/generated_documentation:/tmp/generated_documentation' +
         ' -v $WORKSPACE/docker_doc:/tmp/docker_doc' +
         hgcache_mount_arg +
-        ' doc_task_generation.%s_%s' % (rosdistro_name, doc_repo_spec.name.lower()),
+        ' rosdoc2_task_generation.%s_%s' % (rosdistro_name, doc_repo_spec.name.lower()),
         'echo "# END SECTION"',
     ]),
 ))@
@@ -217,21 +173,18 @@
         '# build and run build_and_install Dockerfile',
         'cd $WORKSPACE/docker_doc',
         'python3 -u $WORKSPACE/ros_buildfarm/scripts/misc/docker_pull_baseimage.py',
-        'docker build --force-rm -t doc.%s_%s .' % (rosdistro_name, doc_repo_spec.name.lower()),
+        'docker build --force-rm -t rosdoc2.%s_%s .' % (rosdistro_name, doc_repo_spec.name.lower()),
         'echo "# END SECTION"',
         '',
         'echo "# BEGIN SECTION: Run Dockerfile - doc"',
         'docker run' +
         ' --rm ' +
         ' --cidfile=$WORKSPACE/docker_doc/docker.cid' +
-        ' -e=TRAVIS=$TRAVIS' +
         ' -v $WORKSPACE/ros_buildfarm:/tmp/ros_buildfarm:ro' +
-        ' -v $WORKSPACE/rosdoc_lite:/tmp/rosdoc_lite:ro' +
-        ' -v $WORKSPACE/catkin-sphinx:/tmp/catkin-sphinx:ro' +
-        ' -v $WORKSPACE/rosdoc_index:/tmp/rosdoc_index:ro' +
+        ' -v $WORKSPACE/rosdoc2:/tmp/rosdoc2:ro' +
         ' -v $WORKSPACE/ws:/tmp/ws' +
         ' -v $WORKSPACE/generated_documentation:/tmp/generated_documentation' +
-        ' doc.%s_%s' % (rosdistro_name, doc_repo_spec.name.lower()),
+        ' rosdoc2.%s_%s' % (rosdistro_name, doc_repo_spec.name.lower()),
         'echo "# END SECTION"',
     ]),
 ))@
@@ -240,39 +193,17 @@
     script='\n'.join([
         'if [ "$skip_cleanup" = "false" ]; then',
         'echo "# BEGIN SECTION: Clean up to save disk space on agents"',
-        'rm -fr ws/build_isolated',
-        'rm -fr ws/devel_isolated',
-        'rm -fr ws/install_isolated',
-        'rm -fr catkin-sphinx',
-        'rm -fr rosdoc_index',
-        'rm -fr rosdoc_lite',
+        'rm -fr rosdoc2',
         'echo "# END SECTION"',
         'fi',
     ]),
 ))@
 @(SNIPPET(
-    'builder_publish-over-ssh',
-    config_name='docs',
-    remote_directory=rosdistro_name,
-    source_files=[
-        'generated_documentation/api/**/manifest.yaml',
-        'generated_documentation/api/**/stamp',
-        'generated_documentation/changelogs/**/*.html',
-        'generated_documentation/symbols/*.tag',
-
-        'generated_documentation/deps/*',
-        'generated_documentation/hashes/*',
-        'generated_documentation/locations/*',
-        'generated_documentation/metapackage_deps/*',
-    ],
-    remove_prefix='generated_documentation',
-))@
-@(SNIPPET(
     'builder_shell',
     script='\n'.join([
-        'if [ -d "$WORKSPACE/generated_documentation/api_rosdoc" ]; then',
+        'if [ -d "$WORKSPACE/ws/docs_output" ]; then',
         '  echo "# BEGIN SECTION: rsync API documentation to server"',
-        '  cd $WORKSPACE/generated_documentation/api_rosdoc',
+        '  cd $WORKSPACE/ws/docs_output',
         '  for pkg_name in $(find . -maxdepth 1 -mindepth 1 -type d); do',
         '    rsync -e ssh --stats -r --delete $pkg_name %s@%s:%s' % \
           (upload_user, upload_host, os.path.join(upload_root, rosdistro_name, 'api')),
```

</details>

**Review by nuclearsandwich — COMMENTED** [review]
My local run of 
```
generate_doc_script.py https://raw.githubusercontent.com/ros-staging/ros_buildfarm_config/staging/index.yaml rolling default rcutils ubuntu focal amd64 | sh
```
is still going but other than a few questions and suggestions this looks very sound. I raised one larger question that I'd like feedback from other maintainers on but I expect that once I finish running the doc script locally that this will be :heavy_check_mark: from me.

**Comment by clalancette** [comment]
> Did we discuss previously whether creating a separate rosdoc2 job template was the recommended course of action.

We did discuss it a while back, and the consensus at the time was that a separate set of templates would keep the logic in the templates down.  That said, we didn't have any thing concrete to look at when we talked about it, so now that we do we may want to revisit that decision.

> I diffed the two and the differences are almost entirely in parameters to snippet functions and not in overall structure

Yeah, that makes a lot of sense because I more-or-less copied the existing doc jobs and then made modifications as necessary for rosdoc2.  That said, I think that when we add more features to this in the future (around building the packages and doing cross-references between packages), the differences between `doc` and `rosdoc2` will be pronounced enough that it makes sense to split them.

**Comment by nuclearsandwich** [comment]
> Note that the first 2 commits in this series are bugfixes, so can be applied with or without the third one.

I'd also really appreciate if these commits either received a dedicated PR which also highlights the addressed issues or as a minimum, a partial squash of this PR branch before a rebase merge that keeps those commits independent for the sake of keeping git-annotate history clear.

**Comment by clalancette** [comment]
> I'd also really appreciate if these commits either received a dedicated PR which also highlights the addressed issues

I'm going to open a separate PR for the true bugfix that is needed in the groovy script.  The other is just a cleanup, so I actually don't mind too much whether we open a PR for it, merge it as part of this one, or drop it entirely.

File: ros_buildfarm/templates/doc/rosdoc2_job.xml.em
Comment (nuclearsandwich) [inline ros_buildfarm/templates/doc/rosdoc2_job.xml.em:186]: As far as I can tell from running this locally, this generated_documentation directory is entirely unused. Can we remove it entirely from the rosdoc2 templates and container mappings or is there some future use for it?

**Comment by clalancette** [comment]
> I'm going to open a separate PR for the true bugfix that is needed in the groovy script.

See #887

File: ros_buildfarm/config/doc_build_file.py
Comment (clalancette) [reply ros_buildfarm/config/doc_build_file.py:23]: No explicit reason, I just added it at the end :).  I can reorder these how you'd like; would you prefer alphabetically?

File: ros_buildfarm/templates/doc/rosdoc2_job.xml.em
Comment (clalancette) [reply ros_buildfarm/templates/doc/rosdoc2_job.xml.em:186]: Yeah, I think you are right that it is entirely unused.  Let me give it a quick whirl on the test buildfarm, but assuming it is unused I'll remove it completely.

Comment (clalancette) [reply ros_buildfarm/templates/doc/rosdoc2_job.xml.em]: Yep, unused.  Removed in c144255.

File: ros_buildfarm/config/doc_build_file.py
Comment (clalancette) [reply ros_buildfarm/config/doc_build_file.py]: Fixed in c144255

File: ros_buildfarm/config/doc_build_file.py
Comment (nuclearsandwich) [reply ros_buildfarm/config/doc_build_file.py:23]: I think putting both the constant definitions and the list of contents in alphabetical ordering is the thing that makes the most sense but I also think that's a discretionary change.

**Review by nuclearsandwich — APPROVED** [review]

**Comment by clalancette** [comment]
I ended up rebasing this onto the latest after merging #887 .  While I was doing that, I squashed the fix commits into the main commit so we can "Rebase and merge" this.  Finally, I also addressed the comment at https://github.com/ros-infrastructure/ros_buildfarm/pull/880#discussion_r641723623 and alphabetized things.

@nuclearsandwich I'd appreciate one more look before I merge.

**Comment by nuclearsandwich** [comment]
Running locally I get the following message but this is not the correct path for generated documentation which appears to be `ws/docs_build/rcutils/output_staging/generated/doxygen/html`

    Generated documentation: /tmp/rosdoc2_test/generated_documentation/api_rosdoc

**Comment by clalancette** [comment]
> Running locally I get the following message but this is not the correct path for generated documentation which appears to be `ws/docs_build/rcutils/output_staging/generated/doxygen/html`
> 
> ```
> Generated documentation: /tmp/rosdoc2_test/generated_documentation/api_rosdoc
> ```

Actually, neither of those is correct :).  By default, rosdoc2 will put the output from its run in `./docs_output`.  In this case, the PWD inside the docker container is set to should be `/tmp/ws` (where the source is checked out to).  So inside the docker container, we should end up with `/tmp/ws/docs_output/<pkg_names>`.

Outside of the docker container, that should map to `$WORKSPACE/ws/docs_output` (see https://github.com/ros-infrastructure/ros_buildfarm/pull/880/files#diff-084cf0236dc5bd1c30910b3327c848cc8f35719b7b1b7f43f18c228b057636a1R182).  And that is what the following `rsync` relies on as well: https://github.com/ros-infrastructure/ros_buildfarm/pull/880/files#diff-084cf0236dc5bd1c30910b3327c848cc8f35719b7b1b7f43f18c228b057636a1R202 .

That sequence also seems to be borne out by my test run of rcutils here (the rsync doesn't succeed because I have bogus credentials in there, but it otherwise looks like I expect): https://build.test.ros2.org/view/Rdoc/job/Rdoc__rcutils__ubuntu_focal_amd64/34/console

Can you give me a bit more information on how you ran this locally so I can try to figure out what happened?

**Comment by nuclearsandwich** [comment]
> Can you give me a bit more information on how you ran this locally so I can try to figure out what happened?

Yeah I'm running with `generate_doc_script.py` using the staging config.
```
 generate_doc_script.py https://raw.githubusercontent.com/ros-staging/ros_buildfarm_config/staging/index.yaml rolling default rcutils ubuntu focal amd64
```

**Comment by clalancette** [comment]
> Yeah I'm running with `generate_doc_script.py` using the staging config.

OK, I see what happened with this invocation.

First, the run of `rosdoc2` with rcutils currently doesn't complete successfully; looking at the output from the sphinx builder at the end, we see:

```
Exception occurred:
  File "/home/buildfarm/.local/lib/python3.8/site-packages/sphinx/domains/cpp.py", line 6271, in _parse_type
    raise self._make_multi_error(prevErrors, header) from exTyped
sphinx.util.cfamily.DefinitionError: Error when parsing function declaration.
If the function has no return type:
  Error in declarator or parameters-and-qualifiers
  Invalid C++ declaration: Expected identifier in nested name, got keyword: typedef [error at 7]
    typedef _Atomic (_Bool) atomic_bool
    -------^
If the function has a return type:
  Invalid C++ declaration: Expected identifier in nested name, got keyword: typedef [error at 7]
    typedef _Atomic (_Bool) atomic_bool
    -------^
The full traceback has been saved in /tmp/sphinx-err-k3u9kfna.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
/tmp/ws/docs_build/rcutils/rcutils/conf.py
Sphinx-build exited with return code '2'
```

That's why there is no `docs_output` directory as I expected.

Second, there was a bug in the `generate_doc_script.py` script that had it always print a hard-coded path.

The first problem is a problem in `rosdoc2` in combination with `rcutils`.  It will need to be addressed to successfully generate documentation for that package, but shouldn't be a blocker here.

The second problem I've fixed in 45d6f4a .

There is actually a third problem here in that the generated script should not print at all if the previous steps failed, but that is a separate bug.

@nuclearsandwich I'd appreciate another look, particularly at what I did in 45d6f4a.

**Comment by nuclearsandwich** [comment]
> First, the run of `rosdoc2` with rcutils currently doesn't complete successfully; looking at the output from the sphinx builder at the end, we see:

Hah I didn't even notice. I just ran it and walked away.

**Review by nuclearsandwich — APPROVED** [review]
Looks like a good initial start. Thanks for iterating during review!

**Comment by clalancette** [comment]
All right, I did one more rebase to squash the last fix in.  I'm now going to rebase and merge.  Thanks for the reviews!

**Pull request merged**
